## Assignment 1 - Extra Credit

### Test whether a gaussian model N($\mu$, $\sigma$) for the age distribution of citibike drivers is a sensible model, or if you can find a better fit with another distribution.

- Use 2 tests: KS, AD, KL, chisq (even though we have not talked about it in detail yet) to do this.

- Test at the Normal and a least one other distributions (e.g. Poisson, or Binomial, or Chisq, Lognormal.......)

Extra credit: Divide your sample geographically: by Borrow + split Manhattan in an Uptown and a Downtown sample (use your discretion to do so, but ZIP code is a good idea) and see if you notice any differences in how the age distribution can be modeled.

### I made this code on local computer as urllib dont work on compute environment. I also developed this notebook using python 3

In [1]:
import pylab as pl
import pandas as pd
import numpy as np
import os
import scipy.stats
import json
import urllib.request as ulr

%pylab inline

Populating the interactive namespace from numpy and matplotlib


In [1]:
datestring = '201503'

In [28]:
df = pd.read_csv(datestring + '-citibike-tripdata.csv')
df.head()

tripduration      starttime       stoptime  start station id  \
0           669  3/1/2015 0:00  3/1/2015 0:11               164   
1           750  3/1/2015 0:01  3/1/2015 0:14               258   
2           663  3/1/2015 0:01  3/1/2015 0:12               497   
3           480  3/1/2015 0:02  3/1/2015 0:10               470   
4          1258  3/1/2015 0:02  3/1/2015 0:23               345   

            start station name  start station latitude  \
0              E 47 St & 2 Ave               40.753231   
1  DeKalb Ave & Vanderbilt Ave               40.689407   
2           E 17 St & Broadway               40.737050   
3              W 20 St & 8 Ave               40.743453   
4              W 13 St & 6 Ave               40.736494   

   start station longitude  end station id            end station name  \
0               -73.970325             477             W 41 St & 8 Ave   
1               -73.968855             436    Hancock St & Bedford Ave   
2               -73.990093             477             W 41 St & 8 Ave   
3               -74.000040             491        E 24 St & Park Ave S   
4               -73.997044             473  Rivington St & Chrystie St   

   end station latitude  end station longitude  bikeid    usertype  \
0             40.756405             -73.990026   21409  Subscriber   
1             40.682166             -73.953990   19397  Subscriber   
2             40.756405             -73.990026   20998    Customer   
3             40.740964             -73.986022   21565  Subscriber   
4             40.721101             -73.991925   14693  Subscriber   

   birth year  gender  
0      1987.0       1  
1      1968.0       1  
2         NaN       0  
3      1983.0       1  
4      1970.0       1

In [29]:
# df is the dataframe where the content of the csv file is stored
df['date'] = pd.to_datetime(df['starttime'])
# note that with dataframes I can refer to variables as dictionary keys, 
# i.e. df['starttime'] or as attributes: df.starttime. 
df.head(3)

tripduration      starttime       stoptime  start station id  \
0           669  3/1/2015 0:00  3/1/2015 0:11               164   
1           750  3/1/2015 0:01  3/1/2015 0:14               258   
2           663  3/1/2015 0:01  3/1/2015 0:12               497   

            start station name  start station latitude  \
0              E 47 St & 2 Ave               40.753231   
1  DeKalb Ave & Vanderbilt Ave               40.689407   
2           E 17 St & Broadway               40.737050   

   start station longitude  end station id          end station name  \
0               -73.970325             477           W 41 St & 8 Ave   
1               -73.968855             436  Hancock St & Bedford Ave   
2               -73.990093             477           W 41 St & 8 Ave   

   end station latitude  end station longitude  bikeid    usertype  \
0             40.756405             -73.990026   21409  Subscriber   
1             40.682166             -73.953990   19397  Subscriber   
2             40.756405             -73.990026   20998    Customer   

   birth year  gender                date  
0      1987.0       1 2015-03-01 00:00:00  
1      1968.0       1 2015-03-01 00:01:00  
2         NaN       0 2015-03-01 00:01:00

In [30]:
df.columns

Index(['tripduration', 'starttime', 'stoptime', 'start station id',
       'start station name', 'start station latitude',
       'start station longitude', 'end station id', 'end station name',
       'end station latitude', 'end station longitude', 'bikeid', 'usertype',
       'birth year', 'gender', 'date'],
      dtype='object')

In [31]:
#df is the dataframe where the content of the csv file is stored
df['age'] = 2015 - df['birth year'][(df['usertype'] == 'Subscriber')]
df['ageM'] = 2015 - df['birth year'][(df['usertype'] == 'Subscriber') & (df['gender'] == 1)]
df['ageF'] = 2015 - df['birth year'][(df['usertype'] == 'Subscriber') & (df['gender'] == 2)]

In [32]:
# dropping NaN values
df['age'].dropna(inplace= True)
df['ageM'].dropna(inplace= True)
df['ageF'].dropna(inplace= True)

In [69]:
df.head(3)

tripduration      starttime       stoptime  start station id  \
0           669  3/1/2015 0:00  3/1/2015 0:11               164   
1           750  3/1/2015 0:01  3/1/2015 0:14               258   
2           663  3/1/2015 0:01  3/1/2015 0:12               497   

            start station name  start station latitude  \
0              E 47 St & 2 Ave               40.753231   
1  DeKalb Ave & Vanderbilt Ave               40.689407   
2           E 17 St & Broadway               40.737050   

   start station longitude  end station id          end station name  \
0               -73.970325             477           W 41 St & 8 Ave   
1               -73.968855             436  Hancock St & Bedford Ave   
2               -73.990093             477           W 41 St & 8 Ave   

   end station latitude  end station longitude  bikeid    usertype  \
0             40.756405             -73.990026   21409  Subscriber   
1             40.682166             -73.953990   19397  Subscriber   
2             40.756405             -73.990026   20998    Customer   

   birth year  gender                date   age  ageM  ageF  
0      1987.0       1 2015-03-01 00:00:00  28.0  28.0   NaN  
1      1968.0       1 2015-03-01 00:01:00  47.0  47.0   NaN  
2         NaN       0 2015-03-01 00:01:00   NaN   NaN   NaN

In [34]:
import urllib.request as ulr

In [55]:
    url = 'https://maps.googleapis.com/maps/api/geocode/json?latlng=' + str(df['start station latitude'][0]) + ',' \
    + str(df['start station longitude'][0]) + '&key=AIzaSyAW3x05Zr8JEQBRqnnXaB3G27RQSfy0Bk4'

In [56]:
url

'https://maps.googleapis.com/maps/api/geocode/json?latlng=40.75323098,-73.97032517&key=AIzaSyAW3x05Zr8JEQBRqnnXaB3G27RQSfy0Bk4'

In [77]:
borough = []

In [78]:
borough

[]

In [79]:
for i in range(0, len(df)):
    url = 'https://maps.googleapis.com/maps/api/geocode/json?latlng=' + str(df['start station latitude'][i]) + ',' \
    + str(df['start station longitude'][i]) + '&key=AIzaSyAW3x05Zr8JEQBRqnnXaB3G27RQSfy0Bk4'
    response = ulr.urlopen(url)
    data = response.read().decode("utf-8")
    data = json.loads(data)
    borough.append(data['results'][0]['address_components'][3]['long_name'])

IndexError: list index out of range

In [80]:
borough

['Manhattan',
 'Brooklyn',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'New York',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'New York',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Brooklyn',
 'Brooklyn',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan',
 'Manhattan

In [83]:
df['borough'] = pd.DataFrame(borough)

In [85]:
df

tripduration        starttime         stoptime  start station id  \
0                669    3/1/2015 0:00    3/1/2015 0:11               164   
1                750    3/1/2015 0:01    3/1/2015 0:14               258   
2                663    3/1/2015 0:01    3/1/2015 0:12               497   
3                480    3/1/2015 0:02    3/1/2015 0:10               470   
4               1258    3/1/2015 0:02    3/1/2015 0:23               345   
5                363    3/1/2015 0:03    3/1/2015 0:09               380   
6                187    3/1/2015 0:04    3/1/2015 0:07               470   
7                501    3/1/2015 0:04    3/1/2015 0:13               168   
8                606    3/1/2015 0:05    3/1/2015 0:15               345   
9               1488    3/1/2015 0:05    3/1/2015 0:30               499   
10              1303    3/1/2015 0:05    3/1/2015 0:27              2017   
11               400    3/1/2015 0:06    3/1/2015 0:13               173   
12               504    3/1/2015 0:06    3/1/2015 0:14               507   
13               522    3/1/2015 0:09    3/1/2015 0:17               247   
14               913    3/1/2015 0:09    3/1/2015 0:24               368   
15               516    3/1/2015 0:10    3/1/2015 0:19               229   
16               335    3/1/2015 0:10    3/1/2015 0:16               168   
17               209    3/1/2015 0:10    3/1/2015 0:14               383   
18               764    3/1/2015 0:11    3/1/2015 0:23               369   
19               671    3/1/2015 0:12    3/1/2015 0:23               161   
20               411    3/1/2015 0:12    3/1/2015 0:19               450   
21               740    3/1/2015 0:13    3/1/2015 0:25               402   
22               288    3/1/2015 0:13    3/1/2015 0:18               151   
23               553    3/1/2015 0:14    3/1/2015 0:23               375   
24               464    3/1/2015 0:14    3/1/2015 0:22               174   
25               893    3/1/2015 0:15    3/1/2015 0:30               347   
26               307    3/1/2015 0:15    3/1/2015 0:21               285   
27               673    3/1/2015 0:16    3/1/2015 0:27               394   
28               292    3/1/2015 0:16    3/1/2015 0:21               472   
29               730    3/1/2015 0:17    3/1/2015 0:29               164   
...              ...              ...              ...               ...   
341796          1255  3/31/2015 23:50    4/1/2015 0:11               485   
341797           445  3/31/2015 23:50  3/31/2015 23:58               492   
341798           576  3/31/2015 23:50    4/1/2015 0:00               504   
341799           330  3/31/2015 23:51  3/31/2015 23:56               545   
341800           337  3/31/2015 23:52  3/31/2015 23:57               345   
341801           263  3/31/2015 23:52  3/31/2015 23:56               250   
341802           155  3/31/2015 23:52  3/31/2015 23:55              2002   
341803           267  3/31/2015 23:53  3/31/2015 23:57               482   
341804           909  3/31/2015 23:53    4/1/2015 0:08               462   
341805           232  3/31/2015 23:53  3/31/2015 23:57               252   
341806           215  3/31/2015 23:53  3/31/2015 23:57               252   
341807           490  3/31/2015 23:54    4/1/2015 0:02               368   
341808          2118  3/31/2015 23:54    4/1/2015 0:29               401   
341809          1016  3/31/2015 23:54    4/1/2015 0:11               457   
341810           269  3/31/2015 23:54  3/31/2015 23:59               347   
341811           443  3/31/2015 23:55    4/1/2015 0:02               379   
341812           380  3/31/2015 23:55    4/1/2015 0:02               466   
341813           533  3/31/2015 23:56    4/1/2015 0:04               492   
341814           809  3/31/2015 23:56    4/1/2015 0:09               152   
341815           275  3/31/2015 23:57    4/1/2015 0:02               293   
341816           289  3/31/2015 23:58    4/1/2015 0:03  

In [87]:
len(df.borough)

341826

In [88]:
df['age_manhattan'] = 2015 - df['birth year'][(df['usertype'] == 'Subscriber')][(df['borough'] == 'Manhattan')]

In [90]:
df['age_manhattan'].dropna(inplace= True)

#### KS Test to Manhattan age distribution (Normal and Poisson)

In [91]:
scipy.stats.kstest(df['age_manhattan'], 'norm', args=(df.age_manhattan.mean(), df.age_manhattan.std()))

KstestResult(statistic=0.10936608089452893, pvalue=0.0)

- Null Hypothesis: the two distributions are identical (the distribution is normal)
- As the p-value (0.00) is < than the desired significance level (let's use 0.05), we can reject the Null Hypothesis. So the age of Citibike customers are not distributed as a normal distribution.

In [95]:
scipy.stats.kstest(df['age_manhattan'], 'poisson', args=(df.age_manhattan.mean(), df.age_manhattan.std()))

KstestResult(statistic=0.5519779740434767, pvalue=0.0)

- Null Hypothesis: the two distributions are identical (the distribution is poisson)
-  As the p-value (0.00) is < than the desired significance level (let's use 0.05), we can reject the Null Hypothesis. So the age of Citibike customers are not distributed as a poisson distribution.

#### AD Test to Manhattan age distribution (Normal)

In [94]:
scipy.stats.anderson(df['age_manhattan'], 'norm')

AndersonResult(statistic=30.596454352118144, critical_values=array([ 0.575,  0.655,  0.786,  0.916,  1.09 ]), significance_level=array([ 15. ,  10. ,   5. ,   2.5,   1. ]))

- null hypothesis: a sample is drawn from a population that follows a particular distribution (in this case, normal)
- As we see, Statistic >> 0.786 (critical value for significance 0.05), so we reject the null hypothesis, i.e. sample is not drawn from a population that follows normal distribution.